plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,238 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_pack.h>
+
+#include "pgsql.h"
+
+void pgsql_conf_destroy(struct flb_pgsql_config *ctx)
+{
+    PGresult *res = NULL;
+
+    if(PQstatus(ctx->conn) == CONNECTION_OK) {
+        while(PQconsumeInput(ctx->conn) == 0) {
+            res = PQgetResult(ctx->conn);
+            if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+                flb_warn("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+            }
+            PQclear(res);
+        }
+    }
+
+    flb_free(ctx->db_hostname);
+    flb_sds_destroy(ctx->db_table);
+    flb_sds_destroy(ctx->timestamp_key);
+
+    PQfinish(ctx->conn);
+
+    free(ctx);
+    ctx = NULL;
+}
+
+static int cb_pgsql_init(struct flb_output_instance *ins,
+                          struct flb_config *config, void *data)
+{
+
+    struct flb_pgsql_config *ctx;
+    const char *conn_str;
+    size_t str_len;
+    PGresult *res;
+    char *query = NULL;
+    char *temp = NULL;
+
+    /* default values just in case */
+
+    if(!ins->host.name) {
+        ins->host.name = flb_strdup(FLB_PGSQL_HOST);
+    }
+
+    if(!ins->host.port) {
+        ins->host.port = FLB_PGSQL_PORT;
+    }
+
+    ctx = flb_calloc(1, sizeof(struct flb_pgsql_config));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+
+    ctx->db_hostname = flb_strdup(ins->host.name);
+    snprintf(ctx->db_port, sizeof(ctx->db_port), "%d", ins->host.port);
+
+
+    ctx->db_name = flb_output_get_property("database", ins);
+    if(!ctx->db_name) {
+        ctx->db_name = FLB_PGSQL_DBNAME;
+    }
+
+    ctx->db_table = flb_sds_create(flb_output_get_property("table", ins));
+    if(!ctx->db_table) {
+        ctx->db_table = flb_sds_create(FLB_PGSQL_TABLE);
+    }
+
+    ctx->db_user = flb_output_get_property("user", ins);
+    if(!ctx->db_user) {
+        flb_warn("[out_pgsql] You didn't supply a valid user to connect, your current unix user will be used");
+    }
+
+    ctx->db_passwd = flb_output_get_property("password", ins);
+    if(!ctx->db_passwd) {
+        flb_warn("[out_pgsql] You didn't supply a password, you should use a password to authenticate against PostgreSQL");
+    }
+
+    ctx->timestamp_key = flb_sds_create(flb_output_get_property("timestamp_key", ins));
+    if(!ctx->timestamp_key) {
+        ctx->timestamp_key = flb_sds_create(FLB_PGSQL_TIMESTAMP_KEY);
```

Comment:
Check that the memory allocation succeeded, here and in line 88 for db_table

Reply:
Yes you're right! but don't know what will fluentbit will do if the init plugin will fail, does this mean that it will not start or it will discard the plugin and keep going?

Reply:
Fluent Bit will quit if it can't init all plugins that are needed. But that's better than it crashes later on because of a null var.

Reply:
That's make sense! thanks for that help =)